README.md
<!-- PROJECT SHIELDS -->
<!--
*** I'm using markdown "reference style" links for readability.
*** Reference links are enclosed in brackets [ ] instead of parentheses ( ).
*** See the bottom of this document for the declaration of the reference variables
*** for contributors-url, forks-url, etc. This is an optional, concise syntax you may use.
*** https://www.markdownguide.org/basic-syntax/#reference-style-links
-->

[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]

<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/SHAARPEC/shaarpec-python-client">
    <img src="images/logo.png" alt="Logo" width="248" height="95">
  </a>

  <p align="center">
    Python client for SHAARPEC Analytics API.
    <br />
    <a href="https://github.com/SHAARPEC/shaarpec-python-client"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://github.com/SHAARPEC/shaarpec-python-client">View Demo</a>
    ·
    <a href="https://github.com/SHAARPEC/shaarpec-python-client/issues">Report Bug</a>
    ·
    <a href="https://github.com/SHAARPEC/shaarpec-python-client/issues">Request Feature</a>
  </p>
</div>

<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#troubleshooting">Troubleshooting</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
  </ol>
</details>

<!-- ABOUT THE PROJECT -->

## About The Project

[![SHAARPEC API screenshot][product-screenshot]](https://www.shaarpec.com)

This is a Python client for simple access to the SHAARPEC Analytics API. Authentication is handled automatically via device flow, authorization code flow, or client credentials flow. Authentication can also be disabled if accessing a public Analytics API.

The SHAARPEC Analytics API provides calculations on the healthcare organization's resources, capacities, clinical outcomes, and much more. These results can be accessed via a standard REST API, which is usually protected by the SHAARPEC Identity Server.

<p align="right">(<a href="#top">back to top</a>)</p>

### Built With

-   [![Httpx][Httpx]][Httpx]
-   [![Python][Python]][Python-url]

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- GETTING STARTED -->

## Getting Started

### Prerequisites

The shaarpec client is used as a standard Python library. It is always a good idea to install the library in a virtual environment.

### Installation

1. Install the library into your virtual environment.
    ```bash
    pip install shaarpec
    ```
2. Store your credentials to the SHAARPEC IdentityServer in an .env file.
    ```bash
    $ cat .env
    OIDCISH_HOST="https://idp.example.com"
    OIDCISH_CLIENT_ID="my client id"
    OIDCISH_CLIENT_SECRET="my client secret"
    OIDCISH_AUDIENCE="shaarpec_api.full_access_scope"
    OIDCISH_SCOPE="openid shaarpec_api.full_access_scope offline_access"
    ```

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- USAGE EXAMPLES -->

## Usage

This library provides a Client class to easily interact with the SHAARPEC Analytics API. The class methods `Client.with_device(...)`, `Client.with_code(...)`, and `Client.with_credentials(...)` create clients that authenticate with the SHAARPEC IdentityServer with either device flow (not tied to an individual user, recommended), code flow (tied to an individual user, for debugging and development), or credentials flow (non-interactive). There is also a class method `Client.without_auth(...)` that does not invoke the IDP server (but will only work if the Analytics API is public, otherwise give 401 Authentication invalid errors).

All API data is returned as `httpx.Response` objects.

Let's look at some code examples on how to get data from the Analytics API. First, import the client.

```python
from shaarpec import Client
```

Next, use [device flow](https://auth0.com/docs/get-started/authentication-and-authorization-flow/device-authorization-flow) or [code flow](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow) to connect the client to the API with the `Client.with_device(...)`, `Client.with_code(...)`, and/or `Client.with_credentials(...)` class methods.

The credentials can either be stored in a .env file in the working directory (as explained in the Prerequisites section), provided as a path, or given directly as arguments to the `auth` dict.

```python
# Create a client with device flow, give authentication details directly.
client = Client.with_device(
        host="https://api.shaarpec.com/",
        auth={
            "host": "https://idp.shaarpec.com",
            "client_id": ...,
            "client_secret": ...,
            "scope": ...,
            "audience": ...
        }
    )
# Create a client with device flow, read authentication details from .env file.
client = Client.with_device(host="https://api.shaarpec.com/", auth="path/to/.env")
```

Here `host` is the base URL to the Analytics API and `auth` is a dictionary with the login credentials. With device flow, the user needs to finish the sign-in by visiting a url provided by the IdentityServer. A message will be shown:

> Visit https://idp.shaarpec.com/device?userCode=XXXXXXXXX to complete sign-in.

The user visits the website, verifies that the user code is correct and confirms the sign-in. After a few seconds, the client will confirm the sign-in:

> SUCCESS: Authentication was successful. Took XX.Y seconds.

The client is now connected to the API. Visit the Analytics API Base URL to interactively test the endpoints and read the documentation about their path and query parameters. These parameters are used in the regular [requests](https://requests.readthedocs.io/en/latest) and [httpx](https://www.python-httpx.org/) way with `client.verb` calls, where `verb` is either `get`, `post`, or `run` (see below).

### GET and POST

The `get` and `post` verbs are supported in the standard way. For example (API responses are returned as `httpx.Response` objects):

```python
client.get("terminology/allergy_type").json()
```

might return

```
{'419263009': 'Allergy to tree pollen',
 '420174000': 'Allergy to wheat',
 '425525006': 'Allergy to dairy product',
 '714035009': 'Allergy to soya',
 '419474003': 'Allergy to mould',
 '232347008': 'Dander (animal) allergy',
 '91934008': 'Allergy to nut',
 '417532002': 'Allergy to fish',
 '300913006': 'Shellfish allergy',
 '232350006': 'House dust mite allergy',
 '418689008': 'Allergy to grass pollen',
 '91935009': 'Allergy to peanuts',
 '91930004': 'Allergy to eggs',
 '300916003': 'Latex allergy',
 '424213003': 'Allergy to bee venom'}
```

or

```python
client.post("population", conditions=["T78.2", "K81.0"])
```

might return

```
[{'patient_origin_id': '4c92f494-3c98-f8dd-1473-da9eb0196f6f',
    'age': '10-16',
    'is_alive': True,
    'gender': 'F',
    'deceased_year': 0},
 ...
]
```

You can send keyword arguments to the underlying `httpx` calls via the `httpx_kwargs` parameter, for example:

```python
client.get("terminology/condition_type/codes", httpx_kwargs={"timeout": 120}).json()
# Sends API call with 120 second timeout)
```

Read about the available parameters in the [httpx documentation](https://www.python-httpx.org/).

### Running tasks

SHAARPEC Analytics API supports long-running tasks by `POST`:ing to `/service/path/to/endpoint`, and then polling with `GET` to `/service/tasks/{task_id}/status` until the result becomes available at `/service/tasks/{task_id}/results`. There is a `run` function in the library that performs this pattern.

For example

```python
task = client.run("population/conditions")
```

will return a task with the comorbidities in the entire population. A task is a [Pydantic](https://docs.pydantic.dev/) model with the following properties:

```python
class Task(BaseModel):
    """A running task."""
    service: str
    task_id: str
    submitted_at: str
    status: str
    success: Optional[bool]
    progress: Optional[float]
    result: Optional[Any]
    error: Optional[Any]
```

As you can see, the success, progress, result and error are optional and updated automatically when available. The method comes with a progress bar which can be disabled via `client.run("path/to/task", progress_bar=False)`. If you want to use the task result in a subsequent command, you can wait (blocking) for the result with the `task.wait_for_result()` method:

```python
task = client.run("path/to/task")
print(f"The result is: {task.wait_for_result()}!")
```

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- TROUBLESHOOTING -->

## Troubleshooting

### Progress bar fails to render in (Dockerized) Jupyter Lab

Sometimes (especially in a Dockerized environment) the progress bar fails to render when running a task. This seems to be related to an issue in `tqdm` and its dependencies. A workaround seems to be to run:

```bash
jupyter lab build
```

in the environment (in a terminal inside the running Docker container) and refresh the browser. This requires a `node.js` installation first, e.g. per:

```
conda install -c conda-forge nodejs
```

See discussion for [issue 1310](https://github.com/tqdm/tqdm/issues/1310)
on the `tqdm` Github repo.

<!-- ROADMAP -->

## Roadmap

See the [open issues](https://github.com/SHAARPEC/shaarpec-python-client/issues) for a full list of proposed features (and known issues).

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- LICENSE -->

## License

Distributed under the MIT License. See `LICENSE` for more information.

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- CONTACT -->

## Contact

SHAARPEC Support - support@shaarpec.com

Project Link: [https://github.com/SHAARPEC/shaarpec-python-client](https://github.com/SHAARPEC/shaarpec-python-client)

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->

[contributors-shield]: https://img.shields.io/github/contributors/SHAARPEC/shaarpec-python-client.svg?style=for-the-badge
[contributors-url]: https://github.com/SHAARPEC/shaarpec-python-client/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/SHAARPEC/shaarpec-python-client.svg?style=for-the-badge
[forks-url]: https://github.com/SHAARPEC/shaarpec-python-client/network/members
[stars-shield]: https://img.shields.io/github/stars/SHAARPEC/shaarpec-python-client.svg?style=for-the-badge
[stars-url]: https://github.com/SHAARPEC/shaarpec-python-client/stargazers
[issues-shield]: https://img.shields.io/github/issues/SHAARPEC/shaarpec-python-client.svg?style=for-the-badge
[issues-url]: https://github.com/SHAARPEC/shaarpec-python-client/issues
[license-shield]: https://img.shields.io/github/license/SHAARPEC/shaarpec-python-client?style=for-the-badge
[license-url]: https://github.com/SHAARPEC/shaarpec-python-client/blob/master/LICENSE
[product-screenshot]: images/screenshot.png
[Httpx]: images/httpx.svg
[Httpx-url]: https://www.python-httpx.org/
[Python]: images/python-3.8.svg
[Python-url]: https://www.python.org/


CHANGELOG.md
## 2.5.0 (July 7, 2023)

-   Add `httpx_kwargs` argument so that API calls can use httpx parameters such as `timeout` etc (#X).
-   Upgrade to oidcish version 1.0.1.

## 2.4.0 (July 6, 2023)

-   Migrated to ruff for linting.
-   Migrated to pydantic v2.
-   Upgraded all dependencies.

## 2.3.3 (May 8, 2023)

-   Fixed progressbar in jupyter again.

## 2.3.2 (May 2, 2023)

-   Fix string format bug when raising failed task error (PR https://github.com/SHAARPEC/shaarpec-python-client/pull/3).

## 2.3.1 (April 6, 2023)

-   Upgrade to latest version of oidcish with fixed refresh flow for client credentials.

## 2.3.0 (March 27, 2023)

-   Add support for client credentials flow.
-   Update dependencies.

## 2.2.2 (February 24, 2023)

-   Fixed package dependency for progressbar.

## 2.2.1 (February 23, 2023)

-   Fixed progressbar in console and notebook (see https://github.com/tqdm/tqdm/issues/1359).

## 2.2.0 (February 22, 2023)

-   Add connect without authentication.
-   Better handling of 401 error.
-   Fix progress bar in jupyter.
-   Upgrade dependencies.

## 2.1.0 (January 12, 2023)

-   Add function that waits for task result.
-   Check that task was accepted before proceeding.

## 2.0.1 (January 5, 2023)

-   Fix and update dependencies.

## 2.0.0 (January 5, 2023)

-   Add `client.run` method to run tasks in the API.
-   Add `client.post` method to POST to API.

## 1.0.1 (August 29, 2022)

-   Add 60 s client timeout by default.
-   Security updates.

## 1.0.0 (July 8, 2022)

-   Initial release.


shaarpec/tasks.py
"""Task module."""
from typing import Optional, Any
import time

from pydantic import BaseModel
from tqdm.auto import tqdm
import background

from shaarpec.utils import Debugger


class Task(BaseModel):
    """A running task."""

    service: str
    task_id: str
    submitted_at: str
    status: str
    success: Optional[bool] = None
    progress: Optional[float] = None
    result: Optional[Any] = None
    error: Optional[Any] = None
    _debugger: Optional[Debugger] = None

    @background.task
    def print(self, update_interval: float = 0.1) -> None:
        """Print the progress of the task."""
        initial = 0 if self.progress is None else int(100 * self.progress)

        with tqdm(initial=initial, total=100) as pbar:
            pbar.set_description(f"Task {self.status}")

            while self.status in ("submitted", "queued", "in_progress"):
                time.sleep(update_interval)
                fraction = 0 if self.progress is None else self.progress
                pbar.update(int(100 * fraction) - pbar.n)
                pbar.set_description(f"Task {self.status}")

            pbar.set_description(
                f"Task completed ({'successfully' if self.success else 'failed'})"
            )

    def wait_for_result(self, poll_interval: float = 0.1) -> Optional[Any]:
        """Return when result becomes available."""
        while self.status in ("submitted", "queued", "in_progress"):
            time.sleep(poll_interval)

        if not self.success:
            raise ValueError(
                f"Task failed with status {self.status} and "
                f"{'error: ' + str(self.error) if self.error else 'no message.'}"
            )

        return self.result


shaarpec/client.py
"""Client for SHAARPEC Analytics API."""
from typing import Optional, Union, Any
from typing_extensions import Self
import pathlib
import time

import background
import httpx

from oidcish import DeviceFlow, CodeFlow, CredentialsFlow

from shaarpec.tasks import Task


class Client:
    """Client for SHAARPEC Analytics API.

    The input is the base URL to the Analytics API and authentication credentials via
    device or code flow. The `.with_device(...)` and `.with_code(...)` class methods
    are provided and can be used to construct a client. See examples of usage.

    API data is returned as `httpx.Response` objects.
    \f
    Examples
    --------
    >>> from shaarpec import Client
    >>> client = Client.with_device(
            host="https://api.shaarpec.com/",
            auth={
                "host": "https://idp.shaarpec.com"
                "client_id": ...,
                "client_secret": ...,
                "audience": ...,
                "scope": ...
            }
        )
    # Or
    # >>> client = Client.with_device(
    #        host="https://api.shaarpec.com/",
    #        auth="./path/to/my/env.file"
    #    )
    #
    >>> client.get("terminology/allergy_type").json()
    {'419263009': 'Allergy to tree pollen',
     '420174000': 'Allergy to wheat',
     '425525006': 'Allergy to dairy product',
     '714035009': 'Allergy to soya',
     '419474003': 'Allergy to mould',
     '232347008': 'Dander (animal) allergy',
     '91934008': 'Allergy to nut',
     '417532002': 'Allergy to fish',
     '300913006': 'Shellfish allergy',
     '232350006': 'House dust mite allergy',
     '418689008': 'Allergy to grass pollen',
     '91935009': 'Allergy to peanuts',
     '91930004': 'Allergy to eggs',
     '300916003': 'Latex allergy',
     '424213003': 'Allergy to bee venom'}
    >>> client.get("population", conditions=["T78.2", "K81.0"])
    [{'patient_origin_id': '4c92f494-3c98-f8dd-1473-da9eb0196f6f',
        'age': '10-16',
        'is_alive': True,
        'gender': 'F',
        'deceased_year': 0},
    ...
    """

    def __init__(
        self,
        host: str,
        auth: Optional[Union[CodeFlow, DeviceFlow, CredentialsFlow]],
        **kwargs,
    ) -> None:
        timeout = kwargs.pop("timeout", 60)
        self._auth = auth
        self._client = httpx.Client(
            base_url=host,
            timeout=timeout,
            headers={"accept": "application/json"},
            **kwargs,
        )

    @classmethod
    def with_device(
        cls, host: str, auth: Union[dict[str, Any], str, None], **kwargs
    ) -> Self:
        """Authenticate with IDP host using device flow.

        The IDP host must support device flow. Authentication can be provided to
        `auth` as a dict, as environment variables, or as the path to an env file. The
        environment variables are always prefixed with OIDCISH, so OIDCISH_CLIENT_ID
        etc.
        \f
        Parameters
        ----------
          host : str
            The Analytics API base URL.
          auth : dict or string
            Authentication details and other arguments.

            If dict, then valid keywords in this dict are:
              host: str, The IDP host name (OIDCISH_HOST).
              client_id: str, The client ID (OIDCISH_CLIENT_ID).
              client_secret: str, The client secret (OIDCISH_CLIENT_SECRET).
              scope: str, A space separated, case-sensitive list of scopes
                (OIDCISH_SCOPE).
                          (default: openid profile offline_access)
              audience: str, The access claim was designated for this audience
                (OIDCISH_AUDIENCE).

            If string, then path to file with the corresponding variables.


        Examples
        --------
        >>> from shaarpec import Client
        >>> client = Client.with_device(
                "https://api.shaarpec.com",
                auth={
                    "host": "https://idp.example.com",
                    "client_id": ...,
                    "client_secret": ...,
                    "scope": ...,
                    "audience": ...,
                }
            )
        # Or, read auth variables from env file
        >>> client = Client.with_device(
                host="https://api.shaarpec.com",
                auth="path/to/my/file.env"
            )
        >>> client.get("terminology/allergy_type").json()
        ...
        """
        match auth:
            case None:
                return cls(host=host, auth=None, **kwargs)
            case str():
                return cls(host=host, auth=DeviceFlow(_env_file=auth), **kwargs)
            case dict():
                auth_host = auth.pop("host")
                return cls(host=host, auth=DeviceFlow(host=auth_host, **auth), **kwargs)
            case _:
                raise TypeError(
                    f"Object {auth} is not of recognized type ({type(auth)})."
                )

    @classmethod
    def with_code(
        cls, host: str, auth: Optional[dict[str, Any]] = None, **kwargs
    ) -> Self:
        """Authenticate with IDP server using code flow.

        The IDP host must support authorization code flow. Authentication can be
        provided to `auth` as a dict, as environment variables, or as the path to an
        env file. The environment variables are always prefixed with OIDCISH, so
        OIDCISH_CLIENT_ID etc.
        \f
        Parameters
        ----------
          host : str
             The Analytics API base URL.
          auth : dict or str,
            Authentication details and other arguments.

            If dict, then valid keyword in this dict are:
              host: str, The IDP host name (OIDCISH_HOST).
              client_id: str, The client ID (OIDCISH_CLIENT_ID).
              client_secret: str, The client secret (OIDCISH_CLIENT_SECRET).
              redirect_uri: str, Must exactly match one of the allowed redirect URIs
                for the client (OIDCISH_REDIRECT_URI).
                (default: http://localhost)
              username: str = The user name (OIDCISH_USERNAME).
              password: str = The user password (OIDCISH_PASSWORD).
              scope: str, A space separated, case-sensitive list of scopes
                (OIDCISH_SCOPE).
                          (default: openid profile offline_access)
              audience: str = The access claim was designated for this audience
                (OIDCISH_AUDIENCE).

        Examples
        --------
        >>> from shaarpec import Client
        >>> client = Client.with_code(
                "https://api.shaarpec.com",
                auth={
                    "host": "https://idp.example.com",
                    "client_id": ...,
                    "client_secret": ...,
                    "redirect_uri": ...,
                    "username": ...,
                    "password": ...,
                    "scope": ...,
                    "audience": ...,
                }
            )
        # Or, read auth variables from env file
        >>> client = Client.with_code(
                host="https://api.shaarpec.com",
                auth="path/to/my/file.env"
            )
        >>> client.get("terminology/allergy_type").json()
        ...
        """
        match auth:
            case None:
                return cls(host=host, auth=None, **kwargs)
            case str():
                return cls(host=host, auth=CodeFlow(_env_file=auth), **kwargs)
            case dict():
                auth_host = auth.pop("host")
                return cls(host=host, auth=CodeFlow(host=auth_host, **auth), **kwargs)
            case _:
                raise TypeError(
                    f"Object {auth} is not of recognized type ({type(auth)})."
                )

    @classmethod
    def with_credentials(
        cls, host: str, auth: Union[dict[str, Any], str, None], **kwargs
    ) -> Self:
        """Authenticate with IDP host using client credentials flow.

        The IDP host must support client credentials flow. Authentication can be
        provided to`auth` as a dict, as environment variables, or as the path to an
        env file. The environment variables are always prefixed with OIDCISH, so
        OIDCISH_CLIENT_ID etc.
        \f
        Parameters
        ----------
          host : str
            The Analytics API base URL.
          auth : dict or string
            Authentication details and other arguments.

            If dict, then valid keywords in this dict are:
              host: str, The IDP host name (OIDCISH_HOST).
              client_id: str, The client ID (OIDCISH_CLIENT_ID).
              client_secret: str, The client secret (OIDCISH_CLIENT_SECRET).
              audience: str, The access claim was designated for this audience
                (OIDCISH_AUDIENCE).

            If string, then path to file with the corresponding variables.

        Examples
        --------
        >>> from shaarpec import Client
        >>> client = Client.with_credentials(
                "https://api.shaarpec.com",
                auth={
                    "host": "https://idp.example.com",
                    "client_id": ...,
                    "client_secret": ...,
                    "audience": ...,
                }
            )
        # Or, read auth variables from env file
        >>> client = Client.with_credentials(
                host="https://api.shaarpec.com",
                auth="path/to/my/file.env"
            )
        >>> client.get("terminology/allergy_type").json()
        ...
        """
        match auth:
            case None:
                return cls(host=host, auth=None, **kwargs)
            case str():
                return cls(host=host, auth=CredentialsFlow(_env_file=auth), **kwargs)
            case dict():
                auth_host = auth.pop("host")
                return cls(
                    host=host, auth=CredentialsFlow(host=auth_host, **auth), **kwargs
                )
            case _:
                raise TypeError(
                    f"Object {auth} is not of recognized type ({type(auth)})."
                )

    @classmethod
    def without_auth(cls, host: str, **kwargs) -> Self:
        """Create a client that does not use authentication.

        \f
        Parameters
        ----------
          host : str
            The Analytics API base URL.

        Examples
        --------
        >>> from shaarpec import Client
        >>> client = Client.without_auth("https://api.shaarpec.com")
        >>> client.get("terminology/allergy_type").json()
        ...
        """
        return cls(host=host, auth=None, **kwargs)

    @property
    def auth(self) -> Optional[Union[CodeFlow, DeviceFlow, CredentialsFlow]]:
        """Return the authentication credentials."""
        return self._auth

    def get(
        self, uri: str, httpx_kwargs: dict[str, Any] = {}, **kwargs
    ) -> httpx.Response:
        """Get the resource at `uri`.

        Keyword arguments are passed as query parameters.
        """
        headers = {}

        if (auth := self.auth) is not None:
            if (credentials := auth.credentials) is not None:
                access_token = credentials.access_token

                headers = {
                    "Authorization": f"Bearer {access_token}",
                    "x-auth-request-access-token": access_token,
                }

        response = self._client.get(
            uri,
            headers=headers,
            params=kwargs,
            **httpx_kwargs,
        )

        return response

    # pylint: disable=too-many-arguments
    def post(
        self,
        uri: str,
        content=None,
        data=None,
        files=None,
        json=None,
        httpx_kwargs: dict[str, Any] = {},
        **kwargs,
    ) -> httpx.Response:
        """Post to the resource at `uri`.

        Keyword arguments are passed as query parameters.
        """
        headers = {}

        if (auth := self.auth) is not None:
            if (credentials := auth.credentials) is not None:
                access_token = credentials.access_token

                headers = {
                    "Authorization": f"Bearer {access_token}",
                    "x-auth-request-access-token": access_token,
                }

        response = self._client.post(
            uri,
            content=content,
            data=data,
            files=files,
            json=json,
            headers=headers,
            params=kwargs,
            **httpx_kwargs,
        )

        return response

    def run(
        self,
        uri: str,
        content=None,
        data=None,
        files=None,
        json=None,
        poll_interval: float = 0.1,
        progress_bar: bool = True,
        httpx_kwargs: dict[str, Any] = {},
        **kwargs,
    ) -> Task:
        """Run a task at resource `uri`.

        A task is equivalent to posting and polling until the job is finished.

        Keyword arguments are passed as query parameters.
        """
        service = pathlib.Path(uri).parts[0]

        response = self.post(
            uri=uri,
            content=content,
            data=data,
            files=files,
            json=json,
            httpx_kwargs=httpx_kwargs,
            **kwargs,
        )

        match response.status_code:
            case 202:
                pass
            case 401:
                raise RuntimeError("401: Not authenticated.")
            case _:
                raise RuntimeError(
                    "Did not receive 202 Accepted for the task. "
                    f"The actual response code was {response.status_code}. "
                    f"The actual message was {response.text}."
                )

        task_id = response.json().get("task_id")
        submitted_at = response.json().get("submitted_at")

        task = Task(
            service=service,
            task_id=task_id,
            submitted_at=submitted_at,
            status="submitted",
            success=None,
            progress=None,
            result=None,
            error=None,
            _debugger=None,
        )

        if progress_bar:
            task.print()

        self._wait_for_task(task, poll_interval=poll_interval, **httpx_kwargs)

        return task

    @background.task
    def _wait_for_task(self, task: Task, poll_interval: float = 0.1, **kwargs) -> None:
        while task.status in ("submitted", "queued", "in_progress"):
            response = self.get(
                f"{task.service}/tasks/{task.task_id}/status", httpx_kwargs=kwargs
            )

            match response.status_code:
                case 401:
                    print(
                        f"Not authorized to run task id {task.task_id} "
                        f"on service {task.service}."
                    )
                    task.success = False
                    task.status = "unauthorized"

                case 404:
                    print(
                        f"Task with id {task.task_id} could not be found "
                        f"on service {task.service}."
                    )
                    task.success = False
                    task.status = "not_found"

                case 200:
                    match response.json():
                        case {"status": "not_found"}:
                            print(
                                f"Task with id {task.task_id} "
                                f"could not be found on service {task.service}."
                            )
                            task.status = "not_found"

                        case {"status": "in_progress" | "queued"} as data:
                            task.progress = data.get("progress")
                            task.status = data.get("status")
                            time.sleep(poll_interval)

                        case {"status": "complete", "success": True} as data:
                            resp = self.get(
                                f"{task.service}/tasks/{task.task_id}/results",
                                httpx_kwargs=kwargs,
                            )

                            task.result = resp.json().get("result")
                            task.progress = 1.0
                            task.success = True
                            task.status = "complete"

                        case _ as data:
                            task.error = data.get("error")
                            task.success = False
                            task.status = "complete"

                case _:
                    print(f"An error occurred on service {task.service}.")
                    task.error = response.json().get("error")
                    task.success = False
                    task.status = "error"
